### DISCUSSION TITLE
Bundle adjustment model or CaNNOLeS optimizer not working?

### DISCUSSION THREAD

**Discussion by ojwoodford** [JuliaSmoothOptimizers/Organization#33] [General]
Hi there. All these packages look interesting, and I wanted to try them out on bundle adjustment. I was expecting the following code to work:

```
using BundleAdjustmentModels, CaNNOLeS
model = BundleAdjustmentModel("problem-16-22106")
stats = cannoles(model)
```

but I get the following error:

```
ERROR: MethodError: no method matching hess_structure_residual!(::BundleAdjustmentModel{Float64, Vector{Float64}}, ::Vector{Int64}, ::Vector{Int64})
Stacktrace:
 [1] hess_structure_residual(nls::BundleAdjustmentModel{Float64, Vector{Float64}})
   @ NLPModels ~/.julia/packages/NLPModels/ZITAG/src/nls/api.jl:286
 [2] solve!(solver::CaNNOLeSSolver{Float64, Vector{Float64}}, nls::BundleAdjustmentModel{Float64, Vector{Float64}}, stats::SolverCore.GenericExecutionStats{Float64, Vector{Float64}, Vector{Float64}, Any}; x::Vector{Float64}, λ::Vector{Float64}, method::Symbol, linsolve::Symbol, max_f::Int64, max_time::Float64, max_inner::Int64, ϵtol::Float64, verbose::Int64, check_small_residual::Bool, always_accept_extrapolation::Bool, δdec::Float64)
   @ CaNNOLeS ~/.julia/packages/CaNNOLeS/EgvjT/src/CaNNOLeS.jl:175
 [3] solve!(solver::CaNNOLeSSolver{Float64, Vector{Float64}}, nls::BundleAdjustmentModel{Float64, Vector{Float64}}, stats::SolverCore.GenericExecutionStats{Float64, Vector{Float64}, Vector{Float64}, Any})
   @ CaNNOLeS ~/.julia/packages/CaNNOLeS/EgvjT/src/CaNNOLeS.jl:114
 [4] solve!(solver::CaNNOLeSSolver{Float64, Vector{Float64}}, model::BundleAdjustmentModel{Float64, Vector{Float64}}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ SolverCore ~/.julia/packages/SolverCore/kzlVQ/src/solver.jl:40
 [5] solve!(solver::CaNNOLeSSolver{Float64, Vector{Float64}}, model::BundleAdjustmentModel{Float64, Vector{Float64}})
   @ SolverCore ~/.julia/packages/SolverCore/kzlVQ/src/solver.jl:38
 [6] cannoles(nls::BundleAdjustmentModel{Float64, Vector{Float64}}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ CaNNOLeS ~/.julia/packages/CaNNOLeS/EgvjT/src/CaNNOLeS.jl:111
 [7] cannoles(nls::BundleAdjustmentModel{Float64, Vector{Float64}})
   @ CaNNOLeS ~/.julia/packages/CaNNOLeS/EgvjT/src/CaNNOLeS.jl:103
```

Am I doing something wrong? Also, can CaNNOLeS optimize robustified least squares costs, or just pure least squares costs?

**Comment by abelsiqueira** [comment]
Hi @ojwoodford, thanks for reporting this. The issue here is that BundleAdjustmentModels doesn't have second-order information on the residual function, and CaNNOLeS tries to use that, by default.
One solution (would be) to use `method = :Newton_noFHess` in the CaNNOLeS call, but I just noticed that it doesn't work.
I created https://github.com/JuliaSmoothOptimizers/CaNNOLeS.jl/issues/110 to track it and I hope to fix it soon.

I don't know the robustified least squares problem. What changes in the optimization problem?

**Reply by abelsiqueira** [reply]
I have released CaNNOLeS 0.7.3 and it should work now with the mentioned option:

[![asciicast](https://asciinema.org/a/561688.svg)](https://asciinema.org/a/561688)

**Reply by ojwoodford** [reply]
Many thanks! I'll take a look.

Robustified least squares is discussed in sections 3.3 and 4.3 of the paper ["Bundle Adjustment — A Modern Synthesis"](https://lear.inrialpes.fr/pubs/2000/TMHF00/Triggs-va99.pdf) (amongst many others).

**Reply by ojwoodford** [reply]
Regarding robustified least squares, do you have any thoughts on whether the algorithm used in CaNNOLeS might still work? Supporting robust costs would make such an algorithm significantly more useful.

**Reply by abelsiqueira** [reply]
I had a quick look, so my understanding might not be accurate.
Short answer is no, CaNNOLeS can't handle it, because the robustified version is specific to Bundle Adjustment models and CaNNOLeS is general.
Slightly longer answer would be that maybe we can modify the models and the solvers to handle this. It looks like it's a fixed change in the model. @AntoninKns, are you still working on the Bundle Adjustment models? Below I explain more.

---

Explanation of what I understood.
The general NLS is to minimize `0.5|r(x)|^2`, which translates to `0.5 * sum(r_j(x)^2 for j = 1:n)`.
In the BA problem, we want to minimize `0.5 * sum(d_i' * W_i * d_i for i = 1:m)`, where `d_i = obs_i - z_i(x)`, `z_i(x)` is a model for the observation `obs_i`, with parameters `x`, and `W_i` is a weights matrix. Note that I used n and m to differentiate the dimensions. To write the BA as a residual function `r(x)` you have to `vcat` the vectors `C_i * d_i`, where `C_i' * C_i = W_i`. I'm guessing obs_i has 3d, so `n = 3m`. 

The robustified version would be to add a `rho_i` **function** for each observation and change the problem to minimize `0.5 * sum(rho_i(d_i' * W_i * d_i) for i = 1:n)`.
Depending on how the BA models are implemented, the `rho_i` could be an option on creation and those could be integrate in the derivatives.

For CaNNOLeS (or other methods) to handle the robustified version directly, there would need to be a way to express the residual function `r(x)`. Is there a way to do this?

Please, correct me if I understood something incorrectly.

**Reply by dpo** [reply]
I agree. The $\rho$ functions would have to be implemented directly in the model.

**Reply by ojwoodford** [reply]
Robustified least squares is very general. I use it most of the NLS problems I optimize, not just bundle adjustment. It's necessary any time the residuals model anything other than Gaussian noise. It's supported by [ceres-solver](http://ceres-solver.org/), a well known and popular C++ NLS library.

If we write the unrobustified problem as (1) `0.5 * sum(|r_i(x)|^2)`, the robustified problem is just (2) `0.5 * sum(rho_i(|r_i(x)|^2))`. This can indeed be added into the model, and optimized using iteratively reweighted least squares (IRLS). The issue is that the `r(x)` that the model should output at each iteration, in order to compute the optimal step, is a weighted residual that does not relate to the cost using either of the above two equations. The cost must therefore be computed separately, using the unweighted residuals in equation (2). This means that the model cannot just output `r(x)` and `J(x)`, but should also output `cost(x)`. This is easy enough to implement, and use something like Gauss-Newton of Levenberg-Marquardt to optimize. But I don't know if it would still work using the algorithm implemented in CaNNOLeS. It would be good if it did, because it seems like your algorithm converges in many fewer iterations than Levenberg-Marquardt.

**Reply by abelsiqueira** [reply]
It looks like a good improvement, @ojwoodford, but there are some parts that I'm missing still to know if it is viable to implement.
It seems like the definition of the problem is `sum(|r_i(x_{S_i}|^2)` where `S_i` is a subset of all indexes and each `r_i` is a vector-valued function. In particular, http://ceres-solver.org/nnls_modeling.html#theory seems to imply that.
Can you confirm if that is the case?

Our solvers and NLS models only look at `r(x)`, but a model implementation, such as the one in BundleAdjustmentModels.jl, can keep it separate internally. This means that things that apply to individual `r_i` need to happen at the model level - or specialized solvers.

On the other hand, things that are applied on top of `r(x)`, can be done at the solver level. For instance, changing the problem from `min 0.5|r(x)|^2` to `min 0.5 rho(|r(x)|^2) + lambda * |x|^2`. It can be part of some solver because it doesn't have assumptions on the structure of `r(x)`. I am not saying it is easy nor that we can implement in CaNNOLeS (in particular because it changes the symmetry of the Newton system, it would more than a few tweaks to make sure that it remains SQD).

Now, the fun part is whether we can (and it makes sense to) construct a `RobustifiedModel` wrapper to go on top of any NLS model and solve that problem for all solvers at once.

**Reply by dpo** [reply]
Yes I guess it is possible to implement a thin modeling layer between the raw model and the solver if we provide the derivatives of the $\rho$ functions.

**Reply by ojwoodford** [reply]
Ceres is written based on the assumption that each group of residuals only depend on a subset of variables. This is for practical implementation purposes. However, this is not a necessary assumption to make. All residuals could easily depend on all variables. This has no effect on robustification.

One necessary assumption to make for robustification is that each rho function (or robustifier) applies to a subset of residuals (a subset can be 1 or more residuals), and that these subsets are mutually exclusive.

I am actually in the process of writing my own NLS optimizer in Julia [here](https://github.com/ojwoodford/NLLSsolver.jl), which closely follows the Ceres way of doing things. It already has support for robust costs. It would be straightforward for me to add the solver described in [your paper](https://link.springer.com/article/10.1007/s10589-020-00201-2) to this library. However, at present I'm not sure if the algorithm can be adapted for use with robustified costs, or not. It is not obvious that it can be. So my question is currently more theoretical than practical.

**Reply by AntoninKns** [reply]
Sorry for taking so long to react. I am currently not adding new features but I am upgrading documentation and solving the latest issues in order to help people to add new features and to make a proper release. 
Concerning the idea of adding robustified NLS it could be a nice idea indeed, but I cannot do it right now.
That being said, if you have any questions about the code I would be glad to answer them.
Here is a [link](https://www.gerad.ca/fr/papers/G-2020-42) to the initial conception of the code and a detailed version of the jacobian matrix.

**Reply by ojwoodford** [reply]
The main change that needs to be made is in the solver. It should not assume that `cost = 0.5 * r(x)' * r(x)`, and instead should call a separate cost function, that each model should supply. The model should supply 3 functions:
- `f(x)`: The robustified cost (scalar, or vector to be summed)
- `r(x)`: The weighted residual vector, length N
- `J(x)`: The weighted Jacobian matrix, NxM, where M is the number of model variables

Then any model can easily robustified.

**Reply by tmigot** [reply]
Hi @ojwoodford ! Currently, most of our NLS API assumes that the cost function is `0.5 * r(x) * r(x)`, but using the function `obj` instead is possible. I am adding it to our modeling packages and will keep you updated when it is implemented in the solvers.

**Comment by ojwoodford** [comment]
Hi again. I can't find any example in the documention of a BundleAdjustmentModel actually being optimized. It would help to add this to the [BundleAdjustmentModels tutorial](https://jso.dev/tutorials/introduction-to-bundleadjustmentmodels/). I'm again having trouble getting a BundleAdjustmentModel to optimize, now using any optimizer. See the following:

```
using BundleAdjustmentModel, CaNNOLeS, JSOSolvers
model = BundleAdjustmentModel("problem-16-22106")
stats = tron(model) 

ERROR: MethodError: no method matching jtprod_residual!(::BundleAdjustmentModel{Float64, Vector{…}}, ::Vector{Float64}, ::Vector{Float64}, ::Vector{Float64})
The function `jtprod_residual!` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  jtprod_residual!(::NLPModelsModifiers.FeasibilityFormNLS, ::AbstractVector, ::AbstractVector, ::AbstractVector)
   @ NLPModelsModifiers ~/.julia/packages/NLPModelsModifiers/zy8cI/src/feasibility-form-nls.jl:465
  jtprod_residual!(::NLPModelsModifiers.FeasibilityResidual, ::AbstractVector, ::AbstractVector, ::AbstractVector)
   @ NLPModelsModifiers ~/.julia/packages/NLPModelsModifiers/zy8cI/src/feasibility-residual.jl:119
  jtprod_residual!(::NLPModelsModifiers.SlackNLSModel, ::AbstractVector, ::AbstractVector, ::AbstractVector)
   @ NLPModelsModifiers ~/.julia/packages/NLPModelsModifiers/zy8cI/src/slack-model.jl:751
  ...

Stacktrace:
 [1] objgrad!(nls::BundleAdjustmentModel{…}, x::Vector{…}, g::Vector{…}, Fx::Vector{…}; recompute::Bool)
   @ NLPModels ~/.julia/packages/NLPModels/Cfwb3/src/nls/api.jl:556
 [2] solve!(solver::TrunkSolverNLS{…}, nlp::BundleAdjustmentModel{…}, stats::SolverCore.GenericExecutionStats{…}; callback::JSOSolvers.var"#79#80", x::Vector{…}, atol::Float64, rtol::Float64, Fatol::Float64, Frtol::Float64, max_eval::Int64, max_iter::Int64, max_time::Float64, verbose::Int64, subsolver_verbose::Int64)
   @ JSOSolvers ~/.julia/packages/JSOSolvers/tX6Sk/src/trunkls.jl:269
 [3] solve!(solver::TrunkSolverNLS{…}, model::BundleAdjustmentModel{…}; kwargs::@Kwargs{…})
   @ SolverCoreNLPModelsExt ~/.julia/packages/SolverCore/UI1dA/ext/SolverCoreNLPModelsExt.jl:60
 [4] trunk(::Val{…}, nlp::BundleAdjustmentModel{…}; x::Vector{…}, bk_max::Int64, monotone::Bool, nm_itmax::Int64, subsolver::Symbol, kwargs::@Kwargs{})
   @ JSOSolvers ~/.julia/packages/JSOSolvers/tX6Sk/src/trunkls.jl:220
 [5] trunk(::Val{:GaussNewton}, nlp::BundleAdjustmentModel{Float64, Vector{Float64}})
   @ JSOSolvers ~/.julia/packages/JSOSolvers/tX6Sk/src/trunkls.jl:203
 [6] trunk(nlp::BundleAdjustmentModel{Float64, Vector{Float64}}; variant::Symbol, kwargs::@Kwargs{})
   @ JSOSolvers ~/.julia/packages/JSOSolvers/tX6Sk/src/trunkls.jl:5
 [7] jso_func()
   @ Main ~/.julia/dev/scratch/compare_bundleadjusters.jl:7
 [8] top-level scope
   @ ~/.julia/dev/scratch/compare_bundleadjusters.jl:35
Some type information was truncated. Use `show(err)` to see complete types.

stats = trunk(model) 

ERROR: MethodError: no method matching jtprod_residual!(::BundleAdjustmentModel{Float64, Vector{…}}, ::Vector{Float64}, ::Vector{Float64}, ::Vector{Float64})
The function `jtprod_residual!` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  jtprod_residual!(::NLPModelsModifiers.FeasibilityFormNLS, ::AbstractVector, ::AbstractVector, ::AbstractVector)
   @ NLPModelsModifiers ~/.julia/packages/NLPModelsModifiers/zy8cI/src/feasibility-form-nls.jl:465
  jtprod_residual!(::NLPModelsModifiers.FeasibilityResidual, ::AbstractVector, ::AbstractVector, ::AbstractVector)
   @ NLPModelsModifiers ~/.julia/packages/NLPModelsModifiers/zy8cI/src/feasibility-residual.jl:119
  jtprod_residual!(::NLPModelsModifiers.SlackNLSModel, ::AbstractVector, ::AbstractVector, ::AbstractVector)
   @ NLPModelsModifiers ~/.julia/packages/NLPModelsModifiers/zy8cI/src/slack-model.jl:751
  ...

Stacktrace:
 [1] objgrad!(nls::BundleAdjustmentModel{…}, x::Vector{…}, g::Vector{…}, Fx::Vector{…}; recompute::Bool)
   @ NLPModels ~/.julia/packages/NLPModels/Cfwb3/src/nls/api.jl:556
 [2] solve!(solver::TrunkSolverNLS{…}, nlp::BundleAdjustmentModel{…}, stats::SolverCore.GenericExecutionStats{…}; callback::JSOSolvers.var"#79#80", x::Vector{…}, atol::Float64, rtol::Float64, Fatol::Float64, Frtol::Float64, max_eval::Int64, max_iter::Int64, max_time::Float64, verbose::Int64, subsolver_verbose::Int64)
   @ JSOSolvers ~/.julia/packages/JSOSolvers/tX6Sk/src/trunkls.jl:269
 [3] solve!(solver::TrunkSolverNLS{…}, model::BundleAdjustmentModel{…}; kwargs::@Kwargs{…})
   @ SolverCoreNLPModelsExt ~/.julia/packages/SolverCore/UI1dA/ext/SolverCoreNLPModelsExt.jl:60
 [4] trunk(::Val{…}, nlp::BundleAdjustmentModel{…}; x::Vector{…}, bk_max::Int64, monotone::Bool, nm_itmax::Int64, subsolver::Symbol, kwargs::@Kwargs{})
   @ JSOSolvers ~/.julia/packages/JSOSolvers/tX6Sk/src/trunkls.jl:220
 [5] trunk(::Val{:GaussNewton}, nlp::BundleAdjustmentModel{Float64, Vector{Float64}})
   @ JSOSolvers ~/.julia/packages/JSOSolvers/tX6Sk/src/trunkls.jl:203
 [6] trunk(nlp::BundleAdjustmentModel{Float64, Vector{Float64}}; variant::Symbol, kwargs::@Kwargs{})
   @ JSOSolvers ~/.julia/packages/JSOSolvers/tX6Sk/src/trunkls.jl:5
 [7] jso_func()
   @ Main ~/.julia/dev/scratch/compare_bundleadjusters.jl:7
 [8] top-level scope
   @ ~/.julia/dev/scratch/compare_bundleadjusters.jl:35
Some type information was truncated. Use `show(err)` to see complete types.

stats = cannoles(model, method=:Newton_noFHess; max_time=1000.0)

┌ Warning: linsolve ma57 not available. Using :ldlfactorizations instead
└ @ CaNNOLeS ~/.julia/packages/CaNNOLeS/XBbAl/src/CaNNOLeS.jl:318
ERROR: AssertionError: Dϕ < 0
Stacktrace:
 [1] line_search(x::Vector{…}, r::Vector{…}, λ::Vector{…}, dx::SubArray{…}, dr::SubArray{…}, dλ::Vector{…}, Fx::Vector{…}, cx::Vector{…}, Jx::SparseMatricesCOO.SparseMatrixCOO{…}, Jcx::SparseMatricesCOO.SparseMatrixCOO{…}, ϕ::CaNNOLeS.var"#ϕ#10", xt::Vector{…}, rt::Vector{…}, λt::Vector{…}, Ft::Vector{…}, ct::Vector{…}, F!::CaNNOLeS.var"#F!#8"{…}, c!::CaNNOLeS.var"#c!#9"{…}, ρ::Float64, δ::Float64, η::Float64, trial_computed::Bool, merit::Symbol, params::CaNNOLeS.ParamCaNNOLeS{…})
   @ CaNNOLeS ~/.julia/packages/CaNNOLeS/XBbAl/src/CaNNOLeS.jl:1085
 [2] solve!(solver::CaNNOLeSSolver{…}, nls::BundleAdjustmentModel{…}, stats::SolverCore.GenericExecutionStats{…}; callback::CaNNOLeS.var"#13#14", x::Vector{…}, λ::Vector{…}, use_initial_multiplier::Bool, max_iter::Int64, max_eval::Int64, max_time::Float64, max_inner::Int64, atol::Float64, rtol::Float64, Fatol::Float64, Frtol::Float64, verbose::Int64, always_accept_extrapolation::Bool, δdec::Float64)
   @ CaNNOLeS ~/.julia/packages/CaNNOLeS/XBbAl/src/CaNNOLeS.jl:672
 [3] solve!
   @ ~/.julia/packages/CaNNOLeS/XBbAl/src/CaNNOLeS.jl:418 [inlined]
 [4] #solve!#1
   @ ~/.julia/packages/SolverCore/UI1dA/ext/SolverCoreNLPModelsExt.jl:60 [inlined]
 [5] cannoles(nls::BundleAdjustmentModel{…}; linsolve::Symbol, method::Symbol, kwargs::@Kwargs{…})
   @ CaNNOLeS ~/.julia/packages/CaNNOLeS/XBbAl/src/CaNNOLeS.jl:415
 [6] cannoles
   @ ~/.julia/packages/CaNNOLeS/XBbAl/src/CaNNOLeS.jl:402 [inlined]
 [7] jso_func()
   @ Main ~/.julia/dev/scratch/compare_bundleadjusters.jl:7
 [8] top-level scope
   @ ~/.julia/dev/scratch/compare_bundleadjusters.jl:35
Some type information was truncated. Use `show(err)` to see complete types.
```

I also find cannoles takes way longer than expected. My own solver takes 3 seconds to optimize the problem, while cannoles seems to be taking over 30s per iteration!
